File: api/_src/modules/identity/tenant.ts
/**
 * Tenant-scoped identity repository and explicit platform override
 * (port of platform_app.modules.identity.tenant).
 *
 * Reads deny cross-tenant access by default. A platform operator may select
 * another tenant only through an explicit, reasoned {@link PlatformTenantOverride},
 * which writes an audit event.
 */
import type { PrismaClient, UserProfile } from "@prisma/client";

import type { PrismaTransaction } from "../../db.js";

type Db = PrismaClient | PrismaTransaction;

export class TenantAccessDeniedError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "TenantAccessDeniedError";
  }
}

export class ProfileNotFoundError extends Error {
  constructor(message = "profile not found") {
    super(message);
    this.name = "ProfileNotFoundError";
  }
}

export interface TenantContext {
  readonly kind: "tenant";
  readonly universityId: string;
  readonly actorAccountId?: string | null;
}

export interface PlatformTenantOverride {
  readonly kind: "override";
  readonly actorAccountId: string;
  readonly targetUniversityId: string;
  readonly reason: string;
}

export function tenantContext(
  universityId: string,
  actorAccountId: string | null = null,
): TenantContext {
  return { kind: "tenant", universityId, actorAccountId };
}

export function platformTenantOverride(
  actorAccountId: string,
  targetUniversityId: string,
  reason: string,
): PlatformTenantOverride {
  if (reason.trim() === "") {
    throw new Error("a platform tenant override requires a reason");
  }
  return { kind: "override", actorAccountId, targetUniversityId, reason };
}

export type TenantScope = TenantContext | PlatformTenantOverride;

export class IdentityRepository {
  constructor(private readonly db: Db) {}

  async listProfiles(scope: TenantScope): Promise<UserProfile[]> {
    const universityId = resolveUniversityId(scope);
    await this.auditOverride(scope, "user_profile", null);
    return this.db.userProfile.findMany({
      where: { universityId },
      orderBy: { id: "asc" },
    });
  }

  async getProfile(profileId: string, scope: TenantScope): Promise<UserProfile> {
    const profile = await this.db.userProfile.findUnique({
      where: { id: profileId },
    });
    if (profile === null) {
      throw new ProfileNotFoundError();
    }
    const universityId = resolveUniversityId(scope);
    if (profile.universityId !== universityId) {
      throw new TenantAccessDeniedError("profile belongs to another university");
    }
    await this.auditOverride(scope, "user_profile", profile.id);
    return profile;
  }

  private async auditOverride(
    scope: TenantScope,
    targetType: string,
    targetId: string | null,
  ): Promise<void> {
    if (scope.kind !== "override") {
      return;
    }
    await this.db.auditEvent.create({
      data: {
        universityId: scope.targetUniversityId,
        actorAccountId: scope.actorAccountId,
        action: "platform.cross_tenant.read",
        targetType,
        targetId,
        reason: scope.reason,
        beforeState: undefined,
        afterState: { selected_university_id: scope.targetUniversityId },
      },
    });
  }
}

function resolveUniversityId(scope: TenantScope): string {
  return scope.kind === "override" ? scope.targetUniversityId : scope.universityId;
}

File: src/platform_app/modules/identity/tenant.py
"""Tenant-scoped identity repository and explicit platform override."""

from dataclasses import dataclass
from uuid import UUID

from sqlalchemy import select
from sqlalchemy.orm import Session

from platform_app.modules.identity.models import AuditEvent, UserProfile


class TenantAccessDeniedError(PermissionError):
    """Raised when a tenant context addresses another tenant's data."""


@dataclass(frozen=True, slots=True)
class TenantContext:
    """Authenticated context for one university tenant."""

    university_id: UUID
    actor_account_id: UUID | None = None


@dataclass(frozen=True, slots=True)
class PlatformTenantOverride:
    """Explicit, attributable selection of a tenant by a platform operator."""

    actor_account_id: UUID
    target_university_id: UUID
    reason: str

    def __post_init__(self) -> None:
        if not self.reason.strip():
            raise ValueError("a platform tenant override requires a reason")


TenantScope = TenantContext | PlatformTenantOverride


class IdentityRepository:
    """Identity reads that deny cross-tenant access by default."""

    def __init__(self, session: Session) -> None:
        self._session = session

    def list_profiles(self, scope: TenantScope) -> list[UserProfile]:
        university_id = self._university_id(scope)
        self._audit_override(scope, "user_profile", None)
        return list(
            self._session.scalars(
                select(UserProfile)
                .where(UserProfile.university_id == university_id)
                .order_by(UserProfile.id)
            )
        )

    def get_profile(self, profile_id: UUID, scope: TenantScope) -> UserProfile:
        profile = self._session.get(UserProfile, profile_id)
        if profile is None:
            raise LookupError("profile not found")

        university_id = self._university_id(scope)
        if profile.university_id != university_id:
            raise TenantAccessDeniedError("profile belongs to another university")

        self._audit_override(scope, "user_profile", profile.id)
        return profile

    @staticmethod
    def _university_id(scope: TenantScope) -> UUID:
        if isinstance(scope, PlatformTenantOverride):
            return scope.target_university_id
        return scope.university_id

    def _audit_override(
        self, scope: TenantScope, target_type: str, target_id: UUID | None
    ) -> None:
        if not isinstance(scope, PlatformTenantOverride):
            return
        self._session.add(
            AuditEvent(
                university_id=scope.target_university_id,
                actor_account_id=scope.actor_account_id,
                action="platform.cross_tenant.read",
                target_type=target_type,
                target_id=target_id,
                reason=scope.reason,
                before_state=None,
                after_state={"selected_university_id": str(scope.target_university_id)},
            )
        )
        self._session.flush()
